File: android/app/src/main/java/cat/plou/alarm/Window.kt
package cat.plou.alarm

import java.time.DayOfWeek
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

/** Franjas horarias locales (silencio y vigilancia) con soporte de zona horaria. */

data class LocalTimeInfo(
    /** Minutos transcurridos desde medianoche. */
    val minutes: Int,
    /** Día de la semana, 0 = domingo. */
    val weekday: Int,
)

/** Hora local y día de la semana en la zona horaria indicada. */
fun localTime(at: Long, timezone: String): LocalTimeInfo {
    // Zona horaria desconocida: se cae a UTC en lugar de fallar.
    val zone = runCatching { ZoneId.of(timezone) }.getOrDefault(ZoneId.of("UTC"))
    val t: ZonedDateTime = Instant.ofEpochMilli(at).atZone(zone)
    val weekday = when (t.dayOfWeek) {
        DayOfWeek.SUNDAY -> 0
        DayOfWeek.MONDAY -> 1
        DayOfWeek.TUESDAY -> 2
        DayOfWeek.WEDNESDAY -> 3
        DayOfWeek.THURSDAY -> 4
        DayOfWeek.FRIDAY -> 5
        DayOfWeek.SATURDAY -> 6
    }
    return LocalTimeInfo(t.hour * 60 + t.minute, weekday)
}

private fun parseHhMm(value: String): Int {
    val parts = value.split(":")
    val h = parts.getOrNull(0)?.toIntOrNull() ?: 0
    val m = parts.getOrNull(1)?.toIntOrNull() ?: 0
    return (h % 24) * 60 + (m % 60)
}

/**
 * ¿El instante `at` cae dentro de la franja `from`–`to`?
 *
 * Las franjas que cruzan la medianoche (p. ej. 22:00–07:00) se tratan
 * correctamente: en ese caso el día que se comprueba es el de *inicio* de la
 * franja, de modo que «noches del viernes» incluye la madrugada del sábado.
 */
fun isWithinWindow(
    at: Long,
    timezone: String,
    from: String,
    to: String,
    days: List<Int> = emptyList(),
): Boolean {
    val (minutes, weekday) = localTime(at, timezone)
    val start = parseHhMm(from)
    val end = parseHhMm(to)
    val restricted = days.isNotEmpty() && days.size < 7

    // Franja vacía: no se aplica nunca.
    if (start == end) return false

    if (start < end) {
        if (minutes < start || minutes >= end) return false
        return if (restricted) days.contains(weekday) else true
    }

    // Cruza medianoche.
    if (minutes >= start) return if (restricted) days.contains(weekday) else true
    if (minutes < end) {
        val startDay = (weekday + 6) % 7
        return if (restricted) days.contains(startDay) else true
    }
    return false
}

File: server/src/alarm/window.ts
/** Franjas horarias locales (silencio y vigilancia) con soporte de zona horaria. */

export interface LocalTime {
  /** Minutos transcurridos desde medianoche. */
  minutes: number;
  /** Día de la semana, 0 = domingo. */
  weekday: number;
}

const WEEKDAY_INDEX: Record<string, number> = {
  Sun: 0,
  Mon: 1,
  Tue: 2,
  Wed: 3,
  Thu: 4,
  Fri: 5,
  Sat: 6,
};

const formatterCache = new Map<string, Intl.DateTimeFormat>();

function formatterFor(timezone: string): Intl.DateTimeFormat {
  let f = formatterCache.get(timezone);
  if (!f) {
    try {
      f = new Intl.DateTimeFormat('en-US', {
        timeZone: timezone,
        hour12: false,
        hour: '2-digit',
        minute: '2-digit',
        weekday: 'short',
      });
    } catch {
      // Zona horaria desconocida: se cae a UTC en lugar de fallar.
      f = new Intl.DateTimeFormat('en-US', {
        timeZone: 'UTC',
        hour12: false,
        hour: '2-digit',
        minute: '2-digit',
        weekday: 'short',
      });
    }
    formatterCache.set(timezone, f);
  }
  return f;
}

/** Hora local y día de la semana en la zona horaria indicada. */
export function localTime(at: number, timezone: string): LocalTime {
  const parts = formatterFor(timezone).formatToParts(new Date(at));
  let hour = 0;
  let minute = 0;
  let weekday = 0;
  for (const p of parts) {
    if (p.type === 'hour') hour = Number.parseInt(p.value, 10) % 24;
    else if (p.type === 'minute') minute = Number.parseInt(p.value, 10);
    else if (p.type === 'weekday') weekday = WEEKDAY_INDEX[p.value] ?? 0;
  }
  return { minutes: hour * 60 + minute, weekday };
}

function parseHhMm(value: string): number {
  const [h = '0', m = '0'] = value.split(':');
  return (Number.parseInt(h, 10) % 24) * 60 + (Number.parseInt(m, 10) % 60);
}

/**
 * ¿El instante `at` cae dentro de la franja `from`–`to`?
 *
 * Las franjas que cruzan la medianoche (p. ej. 22:00–07:00) se tratan
 * correctamente: en ese caso el día que se comprueba es el de *inicio* de la
 * franja, de modo que "noches del viernes" incluye la madrugada del sábado.
 */
export function isWithinWindow(
  at: number,
  timezone: string,
  from: string,
  to: string,
  days: readonly number[] = [],
): boolean {
  const { minutes, weekday } = localTime(at, timezone);
  const start = parseHhMm(from);
  const end = parseHhMm(to);
  const restricted = days.length > 0 && days.length < 7;

  if (start === end) {
    // Franja vacía: no se aplica nunca.
    return false;
  }

  if (start < end) {
    const inRange = minutes >= start && minutes < end;
    if (!inRange) return false;
    return restricted ? days.includes(weekday) : true;
  }

  // Cruza medianoche.
  if (minutes >= start) {
    return restricted ? days.includes(weekday) : true;
  }
  if (minutes < end) {
    const startDay = (weekday + 6) % 7;
    return restricted ? days.includes(startDay) : true;
  }
  return false;
}
